### ISSUE TITLE
Azure Disks volume attach still times out on Kubernetes 1.10

### ISSUE THREAD

**Issue by antoineco** [kubernetes/kubernetes#71344]
**What happened**:

After running a rolling-update on a StatefulSet that uses Azure Disks, part of the replicas remain in a `ContainerCreating` state _forever_, exhibiting the following events:

```
Events:
  Type     Reason                  Age                From                                           Message
  ----     ------                  ----               ----                                           -------
  Warning  FailedMount             34m (x92 over 4h)  kubelet, k8s-xyz-123-vmss000001  Unable to mount volumes for pod "solr-0_default(e81d49a2-ee1f-11e8-9bb4-000d3a43da60)": timeout expired waiting for volumes to attach or mount for pod "default"/"solr-0". list of unmounted volumes=[solrhome]. list of unattached volumes=[solrhome]
```

Restarting the `controller-manager` occasionally unblocks the situation:

```
I1122 10:30:51.712314       1 reconciler.go:281] attacherDetacher.AttachVolume started for volume "pvc-xyz-123" (UniqueName: "kubernetes.io/azure-disk//subscriptions/78f9affd-62e8-4720-a4c2-d433dbfdefe8/resourceGroups/cn1vqys6-refparts25-d1/providers/Microsoft.Compute/disks/k8s-mstr-xyz-123-ku-pv-xyz-123") from node "k8s-xyz-123" 
I1122 10:30:52.135191       1 operation_generator.go:332] AttachVolume.Attach succeeded for volume "pvc-xyz-123" (UniqueName: "kubernetes.io/azure-disk//subscriptions/78f9affd-62e8-4720-a4c2-d433dbfdefe8/resourceGroups/cn1vqys6-refparts25-d1/providers/Microsoft.Compute/disks/k8s-mstr-xyz-123-ku-pvc-xyz-123") from node "k8s-xyz-123"
I1122 10:30:52.135517       1 event.go:218] Event(v1.ObjectReference{Kind:"Pod", Namespace:"default", Name:"solr-0", UID:"e81d49a2-ee1f-11e8-9bb4-000d3a43da60", APIVersion:"v1", ResourceVersion:"2949400", FieldPath:""}): type: 'Normal' reason: 'SuccessfulAttachVolume' AttachVolume.Attach succeeded for volume "pvc-xyz-123"
```

**What you expected to happen**:

Attach/detach operations succeed consistently.

**How to reproduce it (as minimally and precisely as possible)**:

Here is a copy of the manifest of the affected StatefulSet:

<details>
<summary>solr-sts.yaml</summary>

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  creationTimestamp: 2018-11-05T16:08:37Z
  generation: 13
  labels:
    app.kubernetes.io/component: search
    app.kubernetes.io/name: solr
  name: solr
  namespace: default
  resourceVersion: "2981863"
  selfLink: /apis/apps/v1/namespaces/default/statefulsets/solr
  uid: 0ddfec77-e115-11e8-b645-000d3a43da60
spec:
  podManagementPolicy: Parallel
  replicas: 2
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.kubernetes.io/component: search
      app.kubernetes.io/name: solr
  serviceName: solr
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.kubernetes.io/component: search
        app.kubernetes.io/name: solr
    spec:
      containers:
      - image: solr
        imagePullPolicy: IfNotPresent
        name: solr
        ports:
        - containerPort: 8983
          name: http
          protocol: TCP
        resources:
          requests:
            cpu: "1"
            memory: 1Gi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /var/solr
          name: solrhome
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      terminationGracePeriodSeconds: 30
  updateStrategy:
    type: RollingUpdate
  volumeClaimTemplates:
  - metadata:
      creationTimestamp: null
      name: solrhome
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 5Gi
    status:
      phase: Pending
status:
  collisionCount: 0
  currentReplicas: 2
  currentRevision: solr-74d956b6d8
  observedGeneration: 13
  readyReplicas: 2
  replicas: 2
  updateRevision: solr-74d956b6d8
```

</details></br>

And the default StorageClass:

<details>
<summary>sc.yaml</summary>

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"storage.k8s.io/v1beta1","kind":"StorageClass","metadata":{"annotations":{"storageclass.beta.kubernetes.io/is-default-class":"true"},"labels":{"kubernetes.io/cluster-service":"true"},"name":"default","namespace":""},"parameters":{"cachingmode":"None","kind":"Managed","storageaccounttype":"Standard_LRS"},"provisioner":"kubernetes.io/azure-disk"}
    storageclass.beta.kubernetes.io/is-default-class: "true"
  creationTimestamp: 2018-11-05T14:58:55Z
  labels:
    kubernetes.io/cluster-service: "true"
  name: default
  resourceVersion: "265"
  selfLink: /apis/storage.k8s.io/v1/storageclasses/default
  uid: 51307313-e10b-11e8-b9f3-000d3a43da60
parameters:
  cachingmode: None
  kind: Managed
  storageaccounttype: Standard_LRS
provisioner: kubernetes.io/azure-disk
reclaimPolicy: Delete
volumeBindingMode: Immediate
```

</details></br>

**Extra information:**

* Two other kind of Pod events that are regularly associated with the error above:

```
  Disk '/subscriptions/xyz-123/resourceGroups/mycluster/providers/Microsoft.Compute/disks/k8s-mstr-mycluster-kub--pvc-xyz-123' cannot be attached as the disk is already owned by VM '/subscriptions/xyz-123/resourceGroups/mycluster/providers/Microsoft.Compute/virtualMachineScaleSets/k8s-myscaleset2-vmss/virtualMachines/k8s-myscaleset2-vmss_1'.
```

```
  Warning  FailedMount             7m (x72 over 2h)   kubelet, k8s-mycluster-vmss00000x  MountVolume.WaitForAttach failed for volume "pvc-xyz-123" : azureDisk - WaitForAttach failed within timeout node (k8s-mycluster-vmss00000x) diskId:(mycluster-kub-pvc-xyz-123) lun:(2)
```

* kubelet logs of a node where attaching the volume is failing

🗒 [kubelet.log](https://github.com/kubernetes/kubernetes/files/2607870/kubelet.log)

* kube-controller-manager logs in an environment where attaching the volume is failing

🗒 [kube-controller-manager.log](https://github.com/kubernetes/kubernetes/files/2608040/kube-controller-manager.log)

**Environment**:
- Kubernetes version (use `kubectl version`): v1.10.10
- Cloud provider or hardware configuration: Azure (westeurope)
- OS (e.g. from /etc/os-release): 16.04.4 LTS
- Kernel (e.g. `uname -a`): 4.13.0-1018-azure
- Install tools: ACS-Engine
- Others:

/kind bug
/sig storage
/sig azure

@andyzhangx @khenidak this might be of interest for you, since you were involved in #70002, which was supposed to fix attach/detach issues in Azure.

**Comment by andyzhangx** [comment]
@antoineco #70002 fix the attach disk issue in kubelet, while in your case, you restarted the controller manager, and then it works. It looks like  disk volume `solrhome`  is never attached to vm `k8s-xyz-123-vmss000001` successfully before you restart controller manager.

**Comment by antoineco** [comment]
Indeed. For completeness I updated the description and added 2 more type of events that are regularly associated to this issue.

**Comment by andyzhangx** [comment]
@antoineco could you check activity logs in that resource group, the time should be before you restart controller manager, you could check whether there is any disk related issue. Since you have already restarted controller manager, we cannot get all logs before your restart.

**Comment by antoineco** [comment]
@andyzhangx I see a bunch of the following failures in the activity log, but the errors stopped 4h before I gathered the logs and restarted the controller-manager:

<details>
<summary>activity log</summary>

```json
{
    "authorization": {
        "action": "Microsoft.Compute/virtualMachineScaleSets/virtualmachines/write",
        "scope": "/subscriptions/mysubscription/resourceGroups/myenv/providers/Microsoft.Compute/virtualMachineScaleSets/k8s-mycluster-vmss/virtualmachines/0"
    },
    "caller": "fb52e24d-f89a-424c-adc4-0e6dffe642c2",
    "channels": "Operation",
    "claims": {
        "aud": "https://management.core.windows.net/",
        "iss": "https://sts.windows.net/42f7676c-f455-423c-82f6-dc2d99791af7/",
        "iat": "1542866160",
        "nbf": "1542866160",
        "exp": "1542870060",
        "aio": "42RgYDg/IVvp5aeWmfYz4uZYMZenAwA=",
        "appid": "663fa753-1cca-4b75-a77c-217cad9cc645",
        "appidacr": "1",
        "http://schemas.microsoft.com/identity/claims/identityprovider": "https://sts.windows.net/42f7676c-f455-423c-82f6-dc2d99791af7/",
        "http://schemas.microsoft.com/identity/claims/objectidentifier": "fb52e24d-f89a-424c-adc4-0e6dffe642c2",
        "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier": "fb52e24d-f89a-424c-adc4-0e6dffe642c2",
        "http://schemas.microsoft.com/identity/claims/tenantid": "42f7676c-f455-423c-82f6-dc2d99791af7",
        "uti": "LttkBbnBhEm-cjoU7t1tAA",
        "ver": "1.0"
    },
    "correlationId": "c85638fd-dbfa-450b-b151-171d07795a75",
    "description": "",
    "eventDataId": "f7e404d8-219f-4e95-b9cf-a73b4191b193",
    "eventName": {
        "value": "EndRequest",
        "localizedValue": "End request"
    },
    "category": {
        "value": "Administrative",
        "localizedValue": "Administrative"
    },
    "eventTimestamp": "2018-11-22T06:37:02.8229171Z",
    "id": "/subscriptions/mysubscription/resourceGroups/myenv/providers/Microsoft.Compute/virtualMachineScaleSets/k8s-mycluster-vmss/virtualmachines/0/events/f7e404d8-219f-4e95-b9cf-a73b4191b193/ticks/636784654228229171",
    "level": "Error",
    "operationId": "c85638fd-dbfa-450b-b151-171d07795a75",
    "operationName": {
        "value": "Microsoft.Compute/virtualMachineScaleSets/virtualmachines/write",
        "localizedValue": "Updates the properties of a Virtual Machine in a Virtual Machine Scale Set"
    },
    "resourceGroupName": "myenv",
    "resourceProviderName": {
        "value": "Microsoft.Compute",
        "localizedValue": "Microsoft.Compute"
    },
    "resourceType": {
        "value": "Microsoft.Compute/virtualMachineScaleSets/virtualmachines",
        "localizedValue": "Microsoft.Compute/virtualMachineScaleSets/virtualmachines"
    },
    "resourceId": "/subscriptions/mysubscription/resourceGroups/myenv/providers/Microsoft.Compute/virtualMachineScaleSets/k8s-mycluster-vmss/virtualmachines/0",
    "status": {
        "value": "Failed",
        "localizedValue": "Failed"
    },
    "subStatus": {
        "value": "Conflict",
        "localizedValue": "Conflict (HTTP Status Code: 409)"
    },
    "submissionTimestamp": "2018-11-22T06:37:22.179904Z",
    "subscriptionId": "mysubscription",
    "properties": {
        "statusCode": "Conflict",
        "serviceRequestId": "bde1a28a-c6af-41ce-8000-29f9dea9d989",
        "statusMessage": "{\"error\":{\"code\":\"AttachDiskWhileBeingDetached\",\"message\":\"Cannot attach data disk '82e2ab0b-b66c-4bec-9b4c-7ea3214d16c6' to VM 'k8s-mycluster-vmss_0' because the disk is currently being detached or the last detach operation failed. Please wait until the disk is completely detached and then try again or delete/detach the disk explicitly again.\"}}"
    },
    "relatedEvents": []
}
```

</details></br>

**Comment by antoineco** [comment]
@andyzhangx I grabbed the kube-controller-manager logs from an environment where the situation is currently happening. See the updated description.

**Comment by andyzhangx** [comment]
Hi @antoineco would you add my slack: andyzhang @kubernetes workspace?

**Comment by andyzhangx** [comment]
@antoineco I think your issue is here, you are using `RollingUpdate` strategy
```
  updateStrategy:
    type: RollingUpdate
```

According to https://github.com/Azure/AKS/issues/477#issuecomment-441061230, if statefulset replica is using azure disk pvc, pls change the strategy to 'Recreate'

**Comment by antoineco** [comment]
@andyzhangx In our deployment every Pod uses a distinct Azure Disk, so there is not multi-attachment.

I'm still waiting feedback from customers, hoping for this to happen again today.

**Comment by andyzhangx** [comment]
@antoineco I have tried in my testing env with v1.10.10 on vmss nodes by `rollingUpdate`, there could be error like following after disk attached to that VM, the disk detect action in vmss node seems slow sometimes, while in the end it could be recovered:
```
Events:
  Type     Reason                  Age                From                                        Message
  ----     ------                  ----               ----                                        -------
  Normal   Scheduled               1m                 default-scheduler                           Successfully assigned statefulset-azuredisk1-0 to k8s-agentpool-39687990-vmss000000
  Normal   SuccessfulMountVolume   1m                 kubelet, k8s-agentpool-39687990-vmss000000  MountVolume.SetUp succeeded for volume "default-token-cdctc"
  Normal   SuccessfulAttachVolume  38s                attachdetach-controller                     AttachVolume.Attach succeeded for volume "pvc-60d77562-ef0c-11e8-bd30-000d3a01450e"
  Normal   SuccessfulAttachVolume  38s                attachdetach-controller                     AttachVolume.Attach succeeded for volume "pvc-60d6a18c-ef0c-11e8-bd30-000d3a01450e"
  Warning  FailedMount             28s                kubelet, k8s-agentpool-39687990-vmss000000  MountVolume.WaitForAttach failed for volume "pvc-60d77562-ef0c-11e8-bd30-000d3a01450e" : azureDisk - WaitForAttach failed within timeout node (k8s-agentpool-39687990-vmss000000) diskId:(andy-vmss11010-dynamic-pvc-60d77562-ef0c-11e8-bd30-000d3a01450e) lun:(0)
  Warning  FailedMount             25s (x2 over 29s)  kubelet, k8s-agentpool-39687990-vmss000000  MountVolume.WaitForAttach failed for volume "pvc-60d6a18c-ef0c-11e8-bd30-000d3a01450e" : azureDisk - WaitForAttach failed within timeout node (k8s-agentpool-39687990-vmss000000) diskId:(andy-vmss11010-dynamic-pvc-60d6a18c-ef0c-11e8-bd30-000d3a01450e) lun:(1)
  Normal   SuccessfulMountVolume   22s                kubelet, k8s-agentpool-39687990-vmss000000  MountVolume.SetUp succeeded for volume "pvc-60d77562-ef0c-11e8-bd30-000d3a01450e"
  Normal   SuccessfulMountVolume   18s                kubelet, k8s-agentpool-39687990-vmss000000  MountVolume.SetUp succeeded for volume "pvc-60d6a18c-ef0c-11e8-bd30-000d3a01450e"
  Normal   Pulling                 15s                kubelet, k8s-agentpool-39687990-vmss000000  pulling image "nginx:alpine"
  Normal   Pulled                  15s                kubelet, k8s-agentpool-39687990-vmss000000  Successfully pulled image "nginx:alpine"
  Normal   Created                 14s                kubelet, k8s-agentpool-39687990-vmss000000  Created container
  Normal   Started                 14s                kubelet, k8s-agentpool-39687990-vmss000000  Started container
```

**Comment by antoineco** [comment]
@andyzhangx thanks for looking into it. In our case the disk never gets attached (> 1d in ContainerCreating state).

**Comment by andyzhangx** [comment]
Hi @antoineco, I found an regression issue of #70002, will send a PR soon. Thanks for reporting.

**Comment by antoineco** [comment]
@andyzhangx I can still observe this issue after applying this patch. Should I open a new issue or can we reopen that one?
